src/simtools/production_configuration/derive_corsika_limits.py

```diff
@@ -192,101 +144,185 @@ def compute_viewcone(self, loss_fraction):
         astropy.units.Quantity
             Viewcone radius in degrees.
         """
-        # Create an index array to create the mapping for the triggered events
-        triggered_indices = np.arange(len(self.shower_id_triggered))
-
-        # Apply the telescope mask if provided
-        if self.telescope_list is not None:
-            mask = np.array(
-                [
-                    all(tel in event for tel in self.telescope_list)
-                    for event in self.trigger_telescope_list_list
-                ]
-            )
-            shower_id_triggered_filtered = self.shower_id_triggered[mask]
-
-            triggered_indices = triggered_indices[mask]
-        else:
-            shower_id_triggered_filtered = self.shower_id_triggered
-
-        array_azimuth_filtered = self.array_azimuth[triggered_indices]
-        array_altitude_filtered = self.array_altitude[triggered_indices]
-
-        sim_azimuth_filtered = self.shower_sim_azimuth[shower_id_triggered_filtered]
-        sim_altitude_filtered = self.shower_sim_altitude[shower_id_triggered_filtered]
-
-        azimuth_diff = array_azimuth_filtered - sim_azimuth_filtered
-        sim_altitude_rad = sim_altitude_filtered
-        array_altitude_rad = array_altitude_filtered
-
-        x_1 = np.cos(azimuth_diff) * np.cos(sim_altitude_rad)
-        y_1 = np.sin(azimuth_diff) * np.cos(sim_altitude_rad)
-        z_1 = np.sin(sim_altitude_rad)
-        x_2 = x_1 * np.sin(array_altitude_rad) - z_1 * np.cos(array_altitude_rad)
-        y_2 = y_1
-        z_2 = x_1 * np.cos(array_altitude_rad) + z_1 * np.sin(array_altitude_rad)
-        off_angles = np.arctan2(np.sqrt(x_2**2 + y_2**2), z_2) * (180.0 / np.pi)
-
-        angle_bins = np.linspace(off_angles.min(), off_angles.max(), 1000)
-        hist, _ = np.histogram(off_angles, bins=angle_bins)
-
-        upper_bin_edge_value = self._compute_limits(
-            hist, angle_bins, loss_fraction, limit_type="upper"
+        hist, _ = np.histogram(self.triggered_data.angular_distance, bins=self.view_cone_bins)
+        return (
+            self._compute_limits(hist, self.view_cone_bins, loss_fraction, limit_type="upper")
+            * u.deg
         )
-        return upper_bin_edge_value * u.deg
 
-    def _transform_to_shower_coordinates(self):
+    def plot_data(self, lower_energy_limit, upper_radial_distance, viewcone, output_path=None):
```

Comment:
Lots of plotting functionality here. I am wondering if some of this can be moved to `visualize.py`?

Reply:
You have a good point here and I was thinking about it. I didn't do it because:

- `plot_data` requires knowledge about the data structure and the selection of histograms is quite specific to the `LimitCalculator`
- this would only `_create_plot` to generalize, which is not enough for a single module.

Looking through the different places: we clearly need to cleanup and refactor the different plotting parts. I have opened an issue to address this with #1462

Reply:
good, thanks